### PR TITLE
Prevent deploying IPv6 network if Zone has no IPv6 DNS configured

### PR DESCRIPTION
**Pull request by GabrielBrascher**
## Description

If an IPv6 enabled network is created but the Zone hasn't IPv6 DNS1 or DNS2 configured then **dnsmasq** inside the Virtual Router does not start.

This PRs adds validation that allows creating an IPv6 network only if the Zone has at least DNS1 or DNS2 configured.

Fixes: #4157 

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 1]
```diff
@@ -3781,7 +3781,7 @@ public Vlan createVlanAndPublicIpRange(final long zoneId, final long networkId,
             ipv4 = true;
         }
 
-        if (startIPv6 != null) {
+        if (vlanIp6Cidr != null) {
```
Comment (GabrielBrascher) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:3784]: CloudStack uses SLAAC for managing IPv6 ranges, therefore it is not necessary to have an start/end IPv6 address; on the other hand, IPv6 CIDR is mandatory for IPv6 networks.

That is why I changed this and other pieces of code that had start/end ipv6 address as mandatory.

File: server/src/main/java/com/cloud/network/NetworkServiceImpl.java [hunk 2]
```diff
@@ -1195,7 +1195,7 @@ public Network createGuestNetwork(CreateNetworkCmd cmd) throws InsufficientCapac
         if (startIP != null) {
             ipv4 = true;
         }
-        if (startIPv6 != null) {
+        if (isNotBlank(ip6Cidr) && isNotBlank(ip6Gateway)) {
```
Comment (GabrielBrascher) [inline server/src/main/java/com/cloud/network/NetworkServiceImpl.java:1198]: Due to SLAAC implementation, an IPv6 network needs CIDR and Gateway. With the refactored code a network is "marked" as ipv6 if IPv6 CIDR and Gateway are not null.

File: server/src/test/java/com/cloud/network/NetworkModelTest.java [hunk 3]
```diff
@@ -90,6 +91,11 @@
     private static final long PHYSICAL_NETWORK_1_ID = 1L;
     private static final long PHYSICAL_NETWORK_2_ID = 2L;
 
+    private static final String IPV6_CIDR = "fd59:16ba:559b:243d::/64";
```
Comment (GabrielBrascher) [inline server/src/test/java/com/cloud/network/NetworkModelTest.java:94]: Nowadays IPv6 networking support only `/64` CIDR. Added Unit tests covering this.

**Comment by GabrielBrascher** [comment]
Fixed conflict with the master.

**Comment by GabrielBrascher** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/network/NetworkModelImpl.java [hunk 4]
```diff
@@ -2207,14 +2207,8 @@ public boolean isNetworkInlineMode(Network network) {
...
     public void checkIp6Parameters(String startIPv6, String endIPv6, String ip6Gateway, String ip6Cidr) throws InvalidParameterValueException {
-        if (!NetUtils.isValidIp6(startIPv6)) {
-            throw new InvalidParameterValueException("Invalid format for the startIPv6 parameter");
-        }
-        if (!NetUtils.isValidIp6(endIPv6)) {
-            throw new InvalidParameterValueException("Invalid format for the endIPv6 parameter");
-        }
 
-        if (!(ip6Gateway != null && ip6Cidr != null)) {
+        if (org.apache.commons.lang3.StringUtils.isBlank(ip6Gateway) || org.apache.commons.lang3.StringUtils.isBlank(ip6Cidr)) {
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/network/NetworkModelImpl.java:2211]: can you encapsulate this in our local proxy to the StringUtils, so migration efforts will be located in a single file?

**Review by DaanHoogland — COMMENTED** [review]
code looks good and makes functional sense. are integration tests for this sensible/possible? (not playing down on the importance of the unit tests you created)

File: server/src/main/java/com/cloud/network/NetworkModelImpl.java (hunk 4 above)
Comment (GabrielBrascher) [reply server/src/main/java/com/cloud/network/NetworkModelImpl.java]: Fixed @DaanHoogland, code is now using the  `com.cloud.utils.StringUtils.isBlank`. Thanks!

**Comment by GabrielBrascher** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@GabrielBrascher do you still want this before 4.15?

**Comment by GabrielBrascher** [comment]
@DaanHoogland It would be great to have this on 4.15. Despite it being a simple code, this PR can prevent some nasty issues when deploying new Zones with IPv6.

Ping for review: @svenvogel @rhtyd @andrijapanicsb @abhinandanprateek @ravening @sureshanaparti @slavkap @kiwiflyer @weizhouapache @RodrigoDLopez @Pearl1594 and **others**.

**Comment by GabrielBrascher** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by Pearl1594 — COMMENTED** [review]
code LGTM.
But @GabrielBrascher do we need to change the ipv6 check/ verification at https://github.com/apache/cloudstack/blob/b66843ecf26146964a17ae6023c0891ed82ff6c1/server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java#L3415 as well?

**Comment by GabrielBrascher** [comment]
@Pearl1594 you are right. It did not affect this implementation due to all networks that I tested already contained start/end ipv4 address (this is something to change in the future, but so far CloudStack can only deploy networks of IPv4 or IPv4 + IPv6; in the future, it could be nice to allow networks with IPv6 nly.

I added a small commit addressing it, thanks for the heads up.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Review by rhtyd — APPROVED** [review]
LGTM, makes sense but haven't tested it.

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
without regression I think this can go in. It's @GabrielBrascher 's neck ;)

**Comment by DaanHoogland** [comment]
@Pearl1594 are you satisfied with this ?

**Review by Pearl1594 — APPROVED** [review]
LGTM

File: server/src/main/java/com/cloud/network/NetworkModelImpl.java [hunk 5]
```diff
@@ -2207,14 +2207,8 @@ public boolean isNetworkInlineMode(Network network) {
...
-        if (!NetUtils.isValidIp6(startIPv6)) {
-            throw new InvalidParameterValueException("Invalid format for the startIPv6 parameter");
-        }
-        if (!NetUtils.isValidIp6(endIPv6)) {
-            throw new InvalidParameterValueException("Invalid format for the endIPv6 parameter");
-        }
 
-        if (!(ip6Gateway != null && ip6Cidr != null)) {
+        if (StringUtils.isBlank(ip6Gateway) || StringUtils.isBlank(ip6Cidr)) {
             throw new InvalidParameterValueException("ip6Gateway and ip6Cidr should be defined when startIPv6/endIPv6 are passed in");
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/network/NetworkModelImpl.java:2212]: @GabrielBrascher may be this exception msg has to be updated, as it is not sure here that startIPv6/endIPv6 are passed or not, as the checks for startIPv6/endIPv6 are moved after this.

Comment (GabrielBrascher) [reply server/src/main/java/com/cloud/network/NetworkModelImpl.java]: Just pushed a commit addressing your comment. Thanks Suresh.

**Comment by GabrielBrascher** [comment]
@DaanHoogland @rhtyd @Pearl1594 @sureshanaparti looking at the codebase for IPv6 there are still a lot of things to improve.

I will leave some of these parts for future PRs as they require testing and so far I made manual tests specifically for the reported issue https://github.com/apache/cloudstack/issues/4157. This PR is partially improving the IPv6, mostly avoiding some critical deployment failures due to a lack of ipv6 gateway/DNS.

Unfortunately, most of our execution flow regarding IPv6 needs to be refactored, this current PR adds a small contribution on behalf of issue #3569, but mainly focused on fixing https://github.com/apache/cloudstack/issues/4157.

I can re-run a couple of manual tests just for the sake. For the (near) future, I think that we should have some IPv6 Marvin tests. I can look into contributing with it.

**Review by DaanHoogland — APPROVED** [review]

**Comment by DaanHoogland** [comment]
@GabrielBrascher the issue this PR solves is marked for 4.13.2. do you want to have this merged on master anyway?

**Comment by GabrielBrascher** [comment]
@DaanHoogland sounds good. Rebased/cherry-picked respective commits aiming branch 4.13.

**Comment by GabrielBrascher** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
not sure if the ape still knows 4.13 but i think it should
@blueorangutan package

**Comment by GabrielBrascher** [comment]
I am not sure as well. Travis seems to not work. Maybe due to the migration of java 8 to java 11

**Comment by GabrielBrascher** [comment]
@DaanHoogland do you think that it would be better to aim this to 4.15 so we can run tests?

**Comment by DaanHoogland** [comment]
@GabrielBrascher you can go for 4.14 as well (or just master)

**Comment by GabrielBrascher** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Pull request merged**